src/pipeline.rs

```diff
@@ -0,0 +1,136 @@
+use std::{
+    sync::{mpsc::Sender, Arc, Mutex},
+    thread,
+    time::Duration,
+};
+
+use crate::{
+    decoder::Decoder,
+    encoder::Encoder,
+    rtp::{RtpFrame, RtpPacker, RtpParser},
+    state::{Frame, SyncHashMap},
+};
+
+pub struct Pipeline {
+    inputs: SyncHashMap<u32, Arc<PipelineInput>>,
+    outputs: SyncHashMap<u32, Arc<PipelineOutput>>,
+    //queue: LiveQueue,
+    //renderer: Renderer,
+}
+
+#[allow(dead_code)]
+pub struct PipelineOutput {
+    encoder: Encoder,
+    rtp_packer: RtpPacker,
+    transport: Mutex<Sender<bytes::Bytes>>,
+}
+
+pub struct PipelineInput {
+    decoder: Decoder,
+    rtp_parser: RtpParser,
+}
+
+// unpack rtp -> decode -> qeueue -> render -> endcode -> pack rtp
+// this code would be in a separate crate (render in this scenario would also separte crate)
+impl Pipeline {
+    pub fn new() -> Self {
+        Pipeline {
+            inputs: SyncHashMap::new(),
+            outputs: SyncHashMap::new(),
+        }
+    }
+
+    pub fn add_input(&self, input_id: u32) {
+        self.inputs.insert(
+            input_id,
+            Arc::new(PipelineInput {
+                decoder: Decoder::new(),
+                rtp_parser: RtpParser::new(),
+            }),
+        );
+    }
+
+    pub fn add_output(&self, input_id: u32, transport: Sender<bytes::Bytes>) {
+        self.outputs.insert(
+            input_id,
+            Arc::new(PipelineOutput {
+                encoder: Encoder::new(),
+                rtp_packer: RtpPacker::new(),
+                transport: Mutex::new(transport),
+            }),
+        );
+    }
+
+    pub fn push_input_data(&self, input_id: u32, buffer: bytes::Bytes) -> Result<()> {
+        let input = self
+            .inputs
+            .get_cloned(&input_id)
+            .ok_or_else(|| anyhow!("no input with id {}.", input_id))?;
+        input.process_input(buffer)?;
+        //self.queue.enqueue(input_id, frames);
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn on_output_data_received(&self, output_id: u32, frames: Vec<Frame>) {
+        let result = self
+            .outputs
+            .get_cloned(&output_id)
+            .ok_or_else(|| anyhow!("no output with id {}.", output_id))
+            .and_then(|output| output.process_output(frames));
+        if let Err(err) = result {
+            eprintln!("{}", err)
+        }
+    }
+
+    pub fn start(self: &Arc<Self>) {
+        let _pipeline = self.clone();
+        thread::spawn(|| {
+            loop {
+                // probablly sth like this
+                //
+                // let input_frames = pipeline.queue.next();
+                // let pipeline.render.render(output_frames);
+                // for let (output_id, frames) in input_frames {
+                //     self.on_output_data_received(output_id, frames)
+                // }
+                eprintln!("render loop");
+                thread::sleep(Duration::from_millis(1000));
+            }
+        });
```

Comment:
In the future, we might want to implement different queuing strategies, that better fit certain use cases. With this render loop, queues have to produce frames in stable intervals, which might not always be the case. I think that this should be a queue responsibility to "push" frames further in the pipeline.

Reply:
it is in this implementation, I expect that next call to be blocking.

Reply:
I got it wrong. Now I understand what you meant here :) The `.next()` will block the `queue`